compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -149,6 +156,8 @@ class PostTyper extends MacroTransform with InfoTransformer { thisPhase =>
     private def processMemberDef(tree: Tree)(using Context): tree.type = {
       val sym = tree.symbol
       Checking.checkValidOperator(sym)
+      if sym.isClass then
+        VarianceChecker.check(tree, privateVarsSetNonLocally)
```

Comment:
Does this work if we find an illegal access from the companion object, defined *after* the class? IIUC the flow of this phase, we won't explore the `object` before we get here.

A test case would be
```scala
class BoxWithCompanion[+A](value: A) {
  private var cached: A = value // error
  def get: A = cached
}

object BoxWithCompanion {
  def put[A](box: BoxWithCompanion[A], value: A): Unit = {
    box.cached = value
  }
}
```